### ISSUE TITLE
Using of Batch Normalization

### ISSUE THREAD

**Issue by dsy412610** [fchollet/keras#2600]
When I used batch normalization in my code, I found that training accuracy increased quickly but testing accuracy was very low. BTW, I used ImageDataGenerator in my code. Who can tell me why testing accuracy can't increase with training accuracy. Thanks!

**Comment by the-moliver** [comment]
Lots of reasons that could happen, but sounds like over fitting. Describe your data and model in more detail and I can give better advice.

**Comment by dsy412610** [comment]
Thanks you very much! The model I used is Alexnet but my dataset is very small. I chose 15 categories of SUN397 for about 13000 images. And if I don't use ImageDataGenerator, the test accuracy is higher than before.

**Comment by the-moliver** [comment]
Still not enough details. Always post your script and some relevant results when asking "why doesn't my thing work?" kinds of questions. Batch-norm makes training faster, which can also make things over-fit faster. Not sure why ImageDataGenerator isn't helping. Are you using a pre-trained network and fine tuning?

**Comment by dsy412610** [comment]
Thanks for your advice.
I don't use pre-trained work and fine tuning. When not using ImageDataGenerator I get the result as following:
Epoch 1/200
13200/13200 [==============================] - 154s - loss: 2.0965 - acc: 0.3228 - val_loss: 3.1294 - val_acc: 0.2157
Epoch 2/200
13200/13200 [==============================] - 154s - loss: 1.8292 - acc: 0.4288 - val_loss: 3.4186 - val_acc: 0.2110
Epoch 3/200
13200/13200 [==============================] - 154s - loss: 1.6922 - acc: 0.4745 - val_loss: 4.0184 - val_acc: 0.1733
Epoch 4/200
13200/13200 [==============================] - 154s - loss: 1.6021 - acc: 0.5036 - val_loss: 2.2243 - val_acc: 0.2900
Epoch 5/200
13200/13200 [==============================] - 154s - loss: 1.5069 - acc: 0.5300 - val_loss: 3.1750 - val_acc: 0.1687
Epoch 6/200
13200/13200 [==============================] - 154s - loss: 1.4087 - acc: 0.5649 - val_loss: 1.8040 - val_acc: 0.4053
Epoch 7/200
13200/13200 [==============================] - 154s - loss: 1.3416 - acc: 0.5859 - val_loss: 1.9013 - val_acc: 0.3680
Epoch 8/200
13200/13200 [==============================] - 154s - loss: 1.2664 - acc: 0.6133 - val_loss: 2.2006 - val_acc: 0.3450
Epoch 9/200
13200/13200 [==============================] - 154s - loss: 1.2133 - acc: 0.6317 - val_loss: 3.5722 - val_acc: 0.1513
Epoch 10/200
13200/13200 [==============================] - 154s - loss: 1.1344 - acc: 0.6568 - val_loss: 2.1280 - val_acc: 0.3580
Epoch 11/200
13200/13200 [==============================] - 154s - loss: 1.0508 - acc: 0.6849 - val_loss: 2.1532 - val_acc: 0.3730
Epoch 12/200
13200/13200 [==============================] - 154s - loss: 0.9694 - acc: 0.7165 - val_loss: 1.7886 - val_acc: 0.4263
Epoch 13/200
13200/13200 [==============================] - 155s - loss: 0.8970 - acc: 0.7395 - val_loss: 2.0956 - val_acc: 0.3637
It's not as good as the result before using batch-norm.
When using ImageDataGenerator:
Epoch 1/200
13200/13200 [==============================] - 442s - loss: 1.9883 - acc: 0.3561 - val_loss: 3.0205 - val_acc: 0.0667
Epoch 2/200
13200/13200 [==============================] - 381s - loss: 2.1541 - acc: 0.2950 - val_loss: 3.1910 - val_acc: 0.0667
Epoch 3/200
13200/13200 [==============================] - 380s - loss: 2.0350 - acc: 0.3249 - val_loss: 3.2049 - val_acc: 0.0667
Epoch 4/200
13200/13200 [==============================] - 380s - loss: 1.9081 - acc: 0.3682 - val_loss: 3.2523 - val_acc: 0.0667
Epoch 5/200
13200/13200 [==============================] - 381s - loss: 1.9492 - acc: 0.3541 - val_loss: 4.1406 - val_acc: 0.0667
Epoch 6/200
13200/13200 [==============================] - 380s - loss: 1.9116 - acc: 0.3617 - val_loss: 3.5204 - val_acc: 0.0667
Epoch 7/200
13200/13200 [==============================] - 380s - loss: 1.8362 - acc: 0.3883 - val_loss: 3.1995 - val_acc: 0.0623
Epoch 8/200
13200/13200 [==============================] - 380s - loss: 1.8902 - acc: 0.3704 - val_loss: 4.1860 - val_acc: 0.0687
Epoch 9/200
13200/13200 [==============================] - 380s - loss: 1.7970 - acc: 0.3961 - val_loss: 3.3728 - val_acc: 0.0667
Epoch 10/200
13200/13200 [==============================] - 381s - loss: 1.8122 - acc: 0.3947 - val_loss: 3.5477 - val_acc: 0.0860
val_acc nearly unchanged.

**Comment by the-moliver** [comment]
And the script?

**Comment by dsy412610** [comment]
![Uploading 1.png…]()

**Comment by dsy412610** [comment]
The model I used as following:
`# output_length = 55
model.add(Convolution2D(96, 11, 11, border_mode = 'valid', input_shape = (3, 227, 227), subsample = (4, 4)))
model.add(BatchNormalization(axis=1))
model.add(Activation('sigmoid'))
model.add(MaxPooling2D(pool_size = (3, 3), strides = (2, 2), border_mode = 'valid'))

#output_length = 27
model.add(ZeroPadding2D(padding = (2, 2)))
model.add(Convolution2D(256, 5, 5, border_mode = 'valid'))
model.add(BatchNormalization(axis=1))
model.add(Activation('sigmoid'))
#output_length = 13
model.add(MaxPooling2D(pool_size = (3, 3), strides = (2, 2), border_mode = 'valid'))


#output_length = 13
model.add(ZeroPadding2D(padding = (1, 1)))
model.add(Convolution2D(384, 3, 3, border_mode = 'valid'))
model.add(BatchNormalization(axis=1))
model.add(Activation('sigmoid'))

#output_length = 13
model.add(ZeroPadding2D(padding = (1, 1)))
model.add(Convolution2D(384, 3, 3, border_mode = 'valid'))
model.add(BatchNormalization(axis=1))
model.add(Activation('sigmoid'))

#output_length = 13
model.add(ZeroPadding2D(padding = (1, 1)))
model.add(Convolution2D(256, 3, 3, border_mode = 'valid'))
model.add(BatchNormalization(axis=1))
model.add(Activation('sigmoid'))
model.add(MaxPooling2D(pool_size = (3, 3), strides = (2, 2), border_mode = 'valid'))

model.add(Flatten())
model.add(Dense(512, init = 'normal'))
model.add(BatchNormalization(axis=1))
model.add(Activation('sigmoid'))

model.add(Dense(512, init = 'normal'))
model.add(BatchNormalization(axis=1))
model.add(Activation('sigmoid'))

model.add(Dense(nb_classes, init = 'normal'))
model.add(BatchNormalization(axis=1))
model.add(Activation('softmax'))`
Then compile and fit:
`sgd = SGD(lr = 0.5, decay = 1e-6, momentum = 0.9, nesterov = True)
model.compile(loss = 'categorical_crossentropy', optimizer = sgd, metrics=['accuracy'])

X_train = X_train.astype('float32')
X_test = X_test.astype('float32')
X_train /= 255
X_test /= 255

datagen = ImageDataGenerator(
        featurewise_center=False,  
        samplewise_center=False,  
        featurewise_std_normalization=False,  
        samplewise_std_normalization=False,  
        zca_whitening=False,  
        rotation_range=1,  
        width_shift_range=0.1,  
        height_shift_range=0.1,  
        horizontal_flip=True,  
        vertical_flip=False)  

    # compute quantities required for featurewise normalization
    # (std, mean, and principal components if ZCA whitening is applied)
    datagen.fit(X_train)#, rounds=2, augment=True)

    model.fit_generator(datagen.flow(X_train, Y_train, batch_size=batch_size),
                        samples_per_epoch=X_train.shape[0],
                        nb_epoch=nb_epoch, show_accuracy=True,
                        validation_data=(X_test, Y_test),
                        nb_worker=1)`

**Comment by dsy412610** [comment]
I used the model as following:
`model = Sequential()
# output_length = 55
model.add(Convolution2D(96, 11, 11, border_mode = 'valid', input_shape = (3, 227, 227), subsample = (4, 4)))
model.add(BatchNormalization(axis=1))
model.add(Activation('sigmoid'))
model.add(MaxPooling2D(pool_size = (3, 3), strides = (2, 2), border_mode = 'valid'))

#output_length = 27
model.add(ZeroPadding2D(padding = (2, 2)))
model.add(Convolution2D(256, 5, 5, border_mode = 'valid'))
model.add(BatchNormalization(axis=1))
model.add(Activation('sigmoid'))
#output_length = 13
model.add(MaxPooling2D(pool_size = (3, 3), strides = (2, 2), border_mode = 'valid'))


#output_length = 13
model.add(ZeroPadding2D(padding = (1, 1)))
model.add(Convolution2D(384, 3, 3, border_mode = 'valid'))
model.add(BatchNormalization(axis=1))
model.add(Activation('sigmoid'))

#output_length = 13
model.add(ZeroPadding2D(padding = (1, 1)))
model.add(Convolution2D(384, 3, 3, border_mode = 'valid'))
model.add(BatchNormalization(axis=1))
model.add(Activation('sigmoid'))

#output_length = 13
model.add(ZeroPadding2D(padding = (1, 1)))
model.add(Convolution2D(256, 3, 3, border_mode = 'valid'))
model.add(BatchNormalization(axis=1))
model.add(Activation('sigmoid'))
model.add(MaxPooling2D(pool_size = (3, 3), strides = (2, 2), border_mode = 'valid'))

model.add(Flatten())
model.add(Dense(512, init = 'normal'))
model.add(BatchNormalization(axis=1))
model.add(Activation('sigmoid'))

model.add(Dense(512, init = 'normal'))
model.add(BatchNormalization(axis=1))
model.add(Activation('sigmoid'))

model.add(Dense(nb_classes, init = 'normal'))
model.add(BatchNormalization(axis=1))
model.add(Activation('softmax'))

# training 
sgd = SGD(lr = 0.5, decay = 1e-6, momentum = 0.9, nesterov = True)
model.compile(loss = 'categorical_crossentropy', optimizer = sgd, metrics=['accuracy'])

X_train = X_train.astype('float32')
X_test = X_test.astype('float32')
X_train /= 255
X_test /= 255

datagen = ImageDataGenerator(
        featurewise_center=False,  
        samplewise_center=False,  
        featurewise_std_normalization=False,  
        samplewise_std_normalization=False,  
        zca_whitening=False,  
        rotation_range=1, 
        width_shift_range=0.1,  
        height_shift_range=0.1,  
        horizontal_flip=True,  
        vertical_flip=False)  

datagen.fit(X_train)
model.fit_generator(datagen.flow(X_train, Y_train, batch_size=batch_size),
                        samples_per_epoch=X_train.shape[0],
                        nb_epoch=nb_epoch, show_accuracy=True,
                        validation_data=(X_test, Y_test),
                        nb_worker=1)`

**Comment by dsy412610** [comment]
I used the model as following:
`model = Sequential()
model.add(Convolution2D(96, 11, 11, border_mode = 'valid', input_shape = (3, 227, 227), subsample = (4, 4)))
model.add(BatchNormalization(axis=1))
model.add(Activation('sigmoid'))
model.add(MaxPooling2D(pool_size = (3, 3), strides = (2, 2), border_mode = 'valid'))

model.add(ZeroPadding2D(padding = (2, 2)))
model.add(Convolution2D(256, 5, 5, border_mode = 'valid'))
model.add(BatchNormalization(axis=1))
model.add(Activation('sigmoid'))
model.add(MaxPooling2D(pool_size = (3, 3), strides = (2, 2), border_mode = 'valid'))

model.add(ZeroPadding2D(padding = (1, 1)))
model.add(Convolution2D(384, 3, 3, border_mode = 'valid'))
model.add(BatchNormalization(axis=1))
model.add(Activation('sigmoid'))

#output_length = 13
model.add(ZeroPadding2D(padding = (1, 1)))
model.add(Convolution2D(384, 3, 3, border_mode = 'valid'))
model.add(BatchNormalization(axis=1))
model.add(Activation('sigmoid'))

model.add(ZeroPadding2D(padding = (1, 1)))
model.add(Convolution2D(256, 3, 3, border_mode = 'valid'))
model.add(BatchNormalization(axis=1))
model.add(Activation('sigmoid'))
model.add(MaxPooling2D(pool_size = (3, 3), strides = (2, 2), border_mode = 'valid'))

model.add(Flatten())
model.add(Dense(512, init = 'normal'))
model.add(BatchNormalization(axis=1))
model.add(Activation('sigmoid'))

model.add(Dense(512, init = 'normal'))
model.add(BatchNormalization(axis=1))
model.add(Activation('sigmoid'))

model.add(Dense(nb_classes, init = 'normal'))
model.add(BatchNormalization(axis=1))
model.add(Activation('softmax'))

sgd = SGD(lr = 0.5, decay = 1e-6, momentum = 0.9, nesterov = True)
model.compile(loss = 'categorical_crossentropy', optimizer = sgd, metrics=['accuracy'])

X_train = X_train.astype('float32')
X_test = X_test.astype('float32')
X_train /= 255
X_test /= 255

datagen = ImageDataGenerator(
        featurewise_center=False,  # set input mean to 0 over the dataset
        samplewise_center=False,  # set each sample mean to 0
        featurewise_std_normalization=False,  # divide inputs by std of the dataset
        samplewise_std_normalization=False,  # divide each input by its std
        zca_whitening=False,  # apply ZCA whitening
        rotation_range=1,  # randomly rotate images in the range (degrees, 0 to 180)
        width_shift_range=0.1,  # randomly shift images horizontally (fraction of total width)
        height_shift_range=0.1,  # randomly shift images vertically (fraction of total height)
        horizontal_flip=True,  # randomly flip images
        vertical_flip=False)  # randomly flip images

datagen.fit(X_train)#, rounds=2, augment=True)

model.fit_generator(datagen.flow(X_train, Y_train, batch_size=batch_size),
                        samples_per_epoch=X_train.shape[0],
                        nb_epoch=nb_epoch, show_accuracy=True,
                        validation_data=(X_test, Y_test),
                        nb_worker=1)`

**Comment by isaacgerg** [comment]
@the-moliver 

I have found the same issue however i use the same data for validation as i did for training.  I think the issue is related to the fact that batchnormalzation behaves different during training than it does during test.  Given this , mode=0 should give training/testing accuracies that are similar for the same input data.  Would you agree?

**Comment by chushan89** [comment]
Hi, I just found one issue, which is that the learning rate (0.5) you set is too high, you could try to reduce the learning rate first. The lr=e-3 or e-4 is commonly used.
